src/compiler/checker.ts

```diff
@@ -5106,7 +5094,9 @@ namespace ts {
         }
 
         function combineTypeMappers(mapper1: TypeMapper, mapper2: TypeMapper): TypeMapper {
-            return t => instantiateType(mapper1(t), mapper2);
+            const mapper: TypeMapper = t => instantiateType(mapper1(t), mapper2);
+            mapper.mappedTypes = mapper1.mappedTypes;

```

Comment:
In #7138 I had the equivalent of `mapper1.mappedTypes.concat(mapper2.mappedTypes)` which I'm pretty sure is wrong. But it should really be `mapper1.mappedTypes.intersect(mapper2.mappedTypes)`, I think.

Since this is an optimization, I don't think it will make a practical difference, because constructing a pathological case would be *very* difficult.